python/lsst/resources/file.py

```diff
@@ -284,13 +284,39 @@ def transfer_from(
                     pass
 
             if transfer == "move":
-                with transaction.undoWith(
-                    f"move from {local_src}", shutil.move, newFullPath, local_src, copy_function=shutil.copy
-                ):
-                    shutil.move(local_src, newFullPath, copy_function=shutil.copy)
+                # If a rename works we try that since that is guaranteed to
+                # be atomic. If that fails we copy and rename. We do this
+                # in case other processes are trying to move to the same
+                # file and we want the "winner" to not be corrupted.
+                try:
+                    with transaction.undoWith(f"move from {local_src}", os.rename, newFullPath, local_src):
+                        os.rename(local_src, newFullPath)
```

Comment:
If rename fails do we actually need "undo"? I'm thinking of a case, for example, when`newFullPath` already exists for some reason but you can't overwrite it, so rename fill fail, does it make sense to move existing `newFullPath` to `local_src`? Maybe if renme fails, it's safe to assume that destination did not change (or does not exist)?

Reply:
If rename fails the undo won't ever happen. `undoWith` only registers the undo callback if the code block succeeds. This is for the case where the rename works but some time later the transaction has to be rolled back.